### PR TITLE
Add a randomizer to run random challenge

### PR DESCRIPTION
**Pull request by greyli**
- Add a random view to redirect to a random challenge
- Add a random button to the index page and challenge page
- Add tests for randomizer

## UI changes

index page:
image

challenge page:

image

### CODE REVIEW

**Comment by 100gle** [comment]
Aha! a funny feature. I believe that we can do more interesting features based on random mode.

But there are something in this PR can be improved:

- set `target="_self"` to `<a>` tag in randomizer will be better than `target="_blank"`. Personally, as an user, I don't want to open more tabs in my browser 🤣.
- put the random button at front of challenge cards will be better, because some level contains too many challenges and expand the vertical spaces so that user can't know where is random button. By the way, this project support desktop and mobile viewport so you might need to consider the css styles whether work as you expected.

**Comment by greyli** [comment]
Thanks for the review!

> * set `target="_self"` to `<a>` tag in randomizer will be better than `target="_blank"`. Personally, as an user, I don't want to open more tabs in my browser 🤣.

Haha, I copied the file from the GitHub button and forgot to update it...

> * put the random button at front of challenge cards will be better...

Updated. The button will look like this on small screen:

image
image

**Comment by laike9m** [comment]
Thanks for adding this interesting feature!

Some feedbacks
- Can you make the text more explict, e.g. `Random Challenge`
- Can you add a tooltip to the icon "Pick a randon challenge"

File: templates/components/randomizer.html [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    display: inline-flex;
+    place-items: center;
+    /* override picocss default text color style for `<a>` tag */
+    color: unset;
+    width: 20px;
+    height: 20px;
+  }
+</style>
+
+<a href="/random" class="icon">
```
Comment (laike9m) [inline templates/components/randomizer.html:12]: Can you add a comment to describe how this svg is generated? Also, curious why we can't use a simple emoji 🔀 for it (same as in `index.html`)?

**Review by laike9m — COMMENTED** [review]
left a comment

File: templates/components/randomizer.html (hunk 1 above)
Comment (100gle) [reply templates/components/randomizer.html:12]: Emoji is very different  among different browsers while SVG not. I prefer to SVG.🤔

**Comment by laike9m** [comment]
Also, can you move the random logo on the homepage below the challenge
list, on the bottom of the page?

Xiaoyue Lin ***@***.***> 于 2023年11月26日周日 上午2:28写道：

File: templates/components/randomizer.html (hunk 1 above)
Comment (greyli) [reply templates/components/randomizer.html:12]: Yes, will do. The SVG here matches the style of the existing icon:

image

**Comment by greyli** [comment]
> @laike9m Also, can you move the random logo on the homepage below the challenge list, on the bottom of the page?

> @100gle put the random button at front of challenge cards will be better, because some level contains too many challenges and expand the vertical spaces so that user can't know where is random button. By the way, this project support desktop and mobile viewport and uses [Pico.css](https://picocss.com/) to give some default style, so you might need to consider the css styles whether work as you expected.

So, should I move it back to the bottom of the page or not...

File: templates/components/randomizer.html (hunk 1 above)
Comment (greyli) [reply templates/components/randomizer.html:12]: Updated with https://github.com/laike9m/Python-Type-Challenges/pull/53/commits/23f4364d09d0e43a34a7365298cd3cf58fb73dfc

- Add tooltip
- Update button text to "Random Challenge"
- Move the button to the bottom of the page
- Add comment for the source of the SVG

**Comment by 100gle** [comment]
> Also, can you move the random logo on the homepage below the challenge list, on the bottom of the page? 

I give the personal opinion why prefer to put the random button before the challenge list.

> put the random button at front of challenge cards will be better, because some level contains too many challenges and expand the vertical spaces so that user can't know where is random button.

In this recommendation we can set some text with them like: 

```html
<div>
  <p>No ideas? try random choice: </p>
  <div class="random-button">...</div>
  <p>or from shallow to deep?</p>
  <div class="challenges">...</div>
</div>
```

surely, it is up to you. @laike9m

File: views/challenge.py [hunk 2]
```diff
@@ -72,6 +73,12 @@ def run_challenge(self, key: ChallengeKey, user_code: str) -> TypeCheckResult:
         challenge = self.get_challenge(key)
         return self._type_check_with_pyright(user_code, challenge.test_code)
 
+    def get_random_challenge(self) -> dict[str, str]:
+        challenges_groupby_level = self.challenges_groupby_level
```
Comment (laike9m) [inline views/challenge.py:77]: This seems redundant?

File: templates/components/randomizer.html (hunk 1 above)
Comment (laike9m) [reply templates/components/randomizer.html:12]: Thanks. Can we also add a tooltip on the randomizer icon on the challenge page?

**Comment by laike9m** [comment]
> > @laike9m Also, can you move the random logo on the homepage below the challenge list, on the bottom of the page?
> 
> > @100gle put the random button at front of challenge cards will be better, because some level contains too many challenges and expand the vertical spaces so that user can't know where is random button. By the way, this project support desktop and mobile viewport and uses [Pico.css](https://picocss.com/) to give some default style, so you might need to consider the css styles whether work as you expected.
> 
> So, should I move it back to the bottom of the page or not...

I still prefer bottom. So no need to change it back.

**Comment by laike9m** [comment]
Could you sync the latest code? I had a few commits. Sorry for the trouble

**Pull request merged**